packages/storage/compat/task.ts

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { UploadTask } from '../src/task';
+import { UploadTaskSnapshotCompat } from './tasksnapshot';
+import { TaskEvent } from '../src/implementation/taskenums';
+import * as types from '@firebase/storage-types';
+import {
+  StorageObserver,
+  ErrorFn,
+  CompleteFn,
+  Subscribe,
+  Unsubscribe
+} from '../src/implementation/observer';
+import { UploadTaskSnapshot } from '../src/tasksnapshot';
+import { ReferenceCompat } from './reference';
+import { FirebaseStorageError } from '../src/implementation/error';
+
+export class UploadTaskCompat implements types.UploadTask {
+  constructor(
+    private readonly delegate: UploadTask,
+    private readonly ref: ReferenceCompat
+  ) {}
+
+  snapshot = new UploadTaskSnapshotCompat(
+    this.delegate.snapshot,
+    this,
+    this.ref
+  );
+
+  cancel = this.delegate.cancel;
+  catch = this.delegate.catch;
+  pause = this.delegate.pause;
+  resume = this.delegate.resume;
```

Comment:
Reassigning class methods may not be safe, because `this` context will change.
`this.delegate.cancel()` expects `this.delegate` as `this`, but when I call `myUploadTaskCompat.cancel()`, `this` becomes `myUploadTaskCompat` which is a compat instance.

They might work because everything the methods in `UploadTask` expects is also available in `UploadTaskCompat`, but a safer pattern would be:
```
  cancel(...args) {
     return this.delegate.cancel(...args)
  }
```
or using bind to preserve the context

```
   cancel = this.delegate.cancel.bind(this.delegate);
```

Reply:
I see this pattern in other places, where this comment also applies.

Reply:
I think this was the only place where they were methods, I looked around and the other similar places, they seem to be properties. Let me know if I missed something.